### DISCUSSION TITLE
How OSPOs are dealing with Log4j vulnerability issue?

### DISCUSSION THREAD

**Discussion by anajsana** [todogroup/ospology#45] [OSPO Discussions]
Thread Discussion coming from TODO Slack. Some people's identities have been anonymized on request:

## Question
Is any OSPO professional directly dealing with Log4j vulnerability in this channel? 🤔 If so, how are you dealing with this issue.

**Comment by anajsana** [comment]
**Anonymous person 1** 

*For us, it was mostly about assisting with discovery using all the standard tools.  software will have bugs so struts or log4j so teams just need to be informed and given mechanisms to keep up with releases.*

**Comment by anajsana** [comment]
**Anonymous person 2**

*I've been working with our security team to help guide them in their vulnerability remediation process. At first they did not include the OSPO in the planning. Then we found one of their remediation processes that was over targeting non-problematic JAR files — distracting engineers from the work of dealing with the problematic ones. So we’re helping them understand the way in which the attack takes place and how to best detect and remediate. One of the problems with false-alarms is that in our industry we’re not allowed to ignore them. We must provide a response to each alarm. So the OSPO is helping our security team make sure we are getting the right signals so the security team can do their job more effectively.*

**Comment by anajsana** [comment]
**Anonymous person 3**

*My role now is to help with the executive communications for the after-action report. The executives want to know what happened, what we did about it, and what we learned from it. Our security team and devops teams each have part of the story, but our risk people have asked the OSPO to serve as a neutral information source about this. I heard that some of the executives think this was an open source problem, but we have data now. We can see how long it took for developers to deploy the fix (less than a day) vs. how long it’s taking to get all the vendors to do the same, or certify that they are not impacted (weeks in some cases).*

**Comment by anajsana** [comment]
**@jlprat (OSPO at Aiven)**

*One of my team colleagues was involved in fixing it for OpenSearch for example. The work was mostly, monitoring the CVE's and providing patches (update the dependency to the right version) to the upstream project, and backporting it all actively maintained versions. So it's not dealing with the vulnerability itself but working on stopping the propagation of the vulnerable library*

**Comment by anajsana** [comment]
**Anonymous person 4**

*Security and development tooling side of things are pretty smooth about "everyone must move to this commit ref". Which meant we had to do some updating of our own OSPO code. We were a more active part of the process for pushing updates on GitHub, and supporting interactions with the log4j project.*

**Comment by anajsana** [comment]
**@zpao (OSPO at Meta)**

We did essentially what others have described. 

1. Find uses, mostly via inspecting the build graph of our apps and services (we mostly use Buck to build), in combination with our open source code search tooling (https://github.com/facebookincubator/Glean). For the long tail, we had to do some grepping and inspection of deployed services.
2. Fix them
3. Redeploy
4. Rinse & repeat as the scale of the issue changed and initial mitigations were not necessarily enough.

To *Anonymous person 4* 's point, software will have bugs and security issues so having plans for how to mitigate these types of things, then ensuring discovery is easy are going to be the best things to have in place.

**Comment by anajsana** [comment]
**@melbgirl (OSPO at NGINX)**

*Same as above for our internal use of the library. We developed a mitigation script that could be easily applied to any instances of NGINX https://www.nginx.com/blog/mitigating-the-log4j-vulnerability-cve-2021-44228-with-nginx/*

**Comment by mswilson** [comment]
Just one action of many during the early days of this incident, and ongoing: the OSPO at Amazon facilitated (during the weekend) the release of the open source licensed hotpatch described here: https://aws.amazon.com/blogs/opensource/hotpatch-for-apache-log4j/

**Comment by tsteenbe** [comment]
The OSPO at [HERE Technologies](https://github.com/heremaps) used [OSS Review Toolkit](https://github.com/oss-review-toolkit/ort) (ORT) in combination with a small custom developed GitLab pipeline that allows us to do instance scanning of all our version control systems whether GitHub, GitLab, etc.  We normally scan our releases with ORT to do license, security and way of working checks but in this case we wanted to be sure we were able to detect log4j everywhere in our code base and not just in our product and services.

If people are interested in our solution I happy to open-source it...

**Comment by jdorfman** [comment]
For those who don't want to set up Glean try [Sourcegraph Notebooks](https://sourcegraph.com/notebooks/Tm90ZWJvb2s6MQ==).

[Example query](https://sourcegraph.com/search?q=org%5C.apache%5C.logging%5C.log4j%27+2%5C.%28%280%7C1%7C2%7C3%7C4%7C5%7C6%7C7%7C8%7C9%7C10%7C11%7C12%7C13%7C14%7C15%29%28%5C.%5B0-9%5D%2B%29%29%0Alang:gradle+count:all&patternType=regexp)
